Ensure that ClickHouse will see ZooKeeper changes on config reloads

azat opened the pull request:
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Ensure that ClickHouse will see ZooKeeper changes on config reloads

Sometimes ClickHouse may not see ZooKeeper changes in case of multiple ZooKeeper replicas, i.e. if you created znode on one replica and ClickHouse connected to another it may not get the latest state. Issue `sync` before reloading the config to avoid this issues.

This should fix the `test_config_substitutions` test (see [1]).

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__[1_6].html

And also I made this test idempotent, otherwise it fails with `--count` [2].

  [2]: https://s3.amazonaws.com/clickhouse-test-reports/72593/422c42a52d675be4bd9ced067f56b4a3a80c195d/integration_tests_flaky_check__asan_.html

Refs: apache/zookeeper@8ee1732

azat committed:
Add ability to sync ZooKeeper before reloading the config

Otherwise it is not deterministic wether ClickHouse will see this
change, i.e. if you created znode just before `SYSTEM RELOAD CONFIG` it
is not 100% that the ClickHouse will see it.

Refs: apache/zookeeper@8ee17324ceb0fd0ace7d87c3bd095ee20c0466aa
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Enable config_reload_sync_zookeeper for test_config_substitutions

This should fix flakiness of `test_config_substitutions/test.py::test_config_multiple_zk_substitutions` [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__[1_6].html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Read config_reload_sync_zookeeper only from the server config

Previously it had been read from each config, i.e. for users it should
be defined in users.xml/yaml.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #72593 from azat/ClickHouse:config-zk-sync -->
*This is an automated comment for commit 422c42a52d675be4bd9ced067f56b4a3a80c195d with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72593/422c42a52d675be4bd9ced067f56b4a3a80c195d/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

Michicosun reviewed (commented):
According to [zookeeper's doc](https://zookeeper.apache.org/doc/r3.8.4/zookeeperInternals.html#sc_consistency) there is no guarantee that read after sync will get up-to-date result and ticket is not closed: [ZOOKEEPER-1675](https://issues.apache.org/jira/browse/ZOOKEEPER-1675).

I think in the [test](https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__[1_6].html) is more likely race between watch callback execution and config reloading thread.

btw ClickHouse Keeper runs sync through the raft log, so this guarantee is achieved there.

azat commented:
>According to [zookeeper's doc](https://zookeeper.apache.org/doc/r3.8.4/zookeeperInternals.html#sc_consistency) there is no guarantee that read after sync will get up-to-date result and ticket is not closed: [ZOOKEEPER-1675](https://issues.apache.org/jira/browse/ZOOKEEPER-1675).

Thank you for the links! This is interesting, this is used for `ReplicatedMergeTree` and in other places, and I guess that those places expecting correct behavior.

>I think in the [test](https://s3.amazonaws.com/clickhouse-test-reports/71406/af822de0bf896dcb0a28f55889d12611bd38171a/integration_tests__asan__old_analyzer__%5B1_6%5D.html) is more likely race between watch callback execution and config reloading thread.

Hm, when I looked at the test logs I found that it did not find the znode

```
2024.11.27 20:53:47.654937 [ 9 ] {2c7313ba-d275-4ea3-a5b5-1a7e644df330} <Warning> ConfigProcessor: Could not get ZooKeeper node: /background_pool_size
```

So I don't see why this could be because of watches, the client (test.py) just created the `/background_pool_size` (note, previously it did not has `from_zk`, so it should not be cached), and then issued a `SYSTEM RELOAD CONFIG` so at this point server should see this znode, and

alexey-milovidov commented:
What about making it unconditionally?

azat commented:
>What about making it unconditionally?

The setting was added to avoid superior load to ZK in case of frequent reloads, but I'm not sure that this could be a problem.

OK, I will do a patch on top that remove the setting

alexey-milovidov commented:
I think it's no problem at all.

azat committed:
tests: make test_config_substitutions idempotent

Now the following passed:

    $ bin=$(which clickhouse); g=$PWD; docker run --privileged -it --rm --volume=$bin:/clickhouse --volume=$bin:/usr/share/clickhouse_fresh --volume=/dev/null:/usr/bin/clickhouse-odbc-bridge --volume=/dev/null:/usr/share/clickhouse-odbc-bridge_fresh --volume=$g/programs/server:/clickhouse-config --volume=$g:/ClickHouse --volume=$g/docker/test/integration/runner/compose:/compose:ro --volume=clickhouse-integration-tests:/var/lib/docker -e PYTEST_ADDOPTS='-vvv -s --pdb --count 2 test_config_substitutions/test.py' --name ch clickhouse/integration-tests-runner

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Remove config_reload_sync_zookeeper (issue sync unconditionally)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[Integration tests (tsan) [6/6]](https://s3.amazonaws.com/clickhouse-test-reports/72593/195f359325161aa8b7e4cdaa673d227911458de8/integration_tests__tsan__[6_6].html)fail: 14, passed: 525

- https://github.com/ClickHouse/ClickHouse/issues/71049

>[AST fuzzer (debug)](https://s3.amazonaws.com/clickhouse-test-reports/72593/195f359325161aa8b7e4cdaa673d227911458de8/ast_fuzzer__debug_.html)Logical error: 'Inconsistent AST formatting: the query:

- https://github.com/ClickHouse/ClickHouse/issues/62975

azat committed:
tests: allow nested with_replace_config()

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on tests/integration/helpers/cluster.py:
```diff
@@ -4542,6 +4543,20 @@ def _create_odbc_config_file(self):
+    def with_replace_config(self, path, replacement):
+        """Create a copy of existing config (if exists) and revert on leaving the context"""
+        self.exec_in_container(
+            ["bash", "-c", f"test ! -f {path} || mv --no-clobber {path} {path}.bak"]
```
Make sense, done.

Michicosun reviewed (approved):
LGTM

Michicosun commented:
failure in `Stress test (asan)` looks unrelated to the changes.
